feat: features OTCR natively in canary

kokekanon opened the pull request:
# 6/6
client modification :  `https://github.com/mehah/otclient/pull/962`


or use this in terminal 

```lua
ProtocolGame.registerOpcode(0x43, function(protocol, msg)
    local count = msg:getU16()
    for i = 1, count do
      local feature = msg:getU8()
	  local enable = msg:getU8()
	  if enable == 1 then
		g_game.enableFeature(feature)
	  end
	end
end)
```


# Description

| Feature                    	| Done | No break v8 or cipsoft 	|
|----------------------------	|-------	|------------------------	|
| 1.- Creature AE + Shader       	|       ✅	|          ✅              	|
| 2.- send disableFeature / enableFeature . easy to modify in config.lua        	|      ✅ https://github.com/mehah/otclient/pull/962 	|                ✅        	|
| 3.- outfit (wings,aura,shader) 	|       ✅	|        ✅                	|
| 4.- TypingIcon                 	|     ✅  	|           ✅             	|
| 5.- map shader support         	|      ✅ 	|         ✅               	|
| 6- item Shader                	|       ✅	|             ✅           	|


## 1.-Creature AE + Shader 
no break cipsoft or v8
![image](https://github.com/user-attachments/assets/5cfcd038-1d37-4c7f-8ed9-ae508e01a7c5)
![img](https://github.com/Nottinghster/otclient/assets/114332266/782e0fcf-b1cf-451e-b102-d7e7943bd50b)
<img src="https://github.com/kokekanon/OTredemption-Picture-NODELETE/blob/main/Picture/Shader/Framebuffer/001_creature.gif?raw=true" width="200" alt="Haskanoid Video" style="max-width:200px;">

## 2.-send disableFeature / enableFeature . easy to modify in config.lua
it is easier for the server administrator to modify config.lua than to compile every time he wants to add or remove a feature.
note1:  v8 and otcr send the same packets, so it could be only 1 function, but i don't want to break anything.

note2 : In otcr requires modification in the client * is not yet ready

![image](https://github.com/user-attachments/assets/55abf9f4-d0a3-4c38-a305-848f0ab8ea4d)


## 3.- outfit (wings,aura,shader)
![image](https://github.com/user-attachments/assets/5edb2aab-e1ad-4975-990e-3c6fce9c00db)

## 4.- TypingIcon
 <img src="https://github.com/kokekanon/OTredemption-Picture-NODELETE/blob/main/Picture/typing%20ico/001.gif?raw=true" width="200" alt="Haskanoid Video" style="max-width:200px;"> 

## 5.- map shader support 
<img src="https://github.com/kokekanon/OTredemption-Picture-NODELETE/blob/main/Picture/Shader/003_map.gif?raw=true" width="200" alt="Peoplemon by Alex Stuart" style="max-width: 200px;">

## 6.- item Shader 
<img src="https://github.com/kokekanon/OTredemption-Picture-NODELETE/blob/main/Picture/Shader/Framebuffer/002_items.gif?raw=true" width="200" alt="Peoplemon by Alex Stuart" style="max-width: 200px;">



## Type of change


  - [x] New feature (non-breaking change which adds functionality)


## How Has This Been Tested
1/6  Creature : AttachedEffect | Shader

```lua
Client:
--   g_game.enableFeature(GameCreatureAttachedEffect)
--   g_game.enableFeature(GameCreatureShader)
Server:
	player:setShader("Outfit - Rainbow")
	player:attachEffectById(7)
	player:attachEffectById(8)
```
2/6  Player: Map Shader
```lua
	player:setMapShader("Map - Party")
```
4/6 feature: Item Shader
```lua
--Client:
	g_game.enableFeature(GameItemShader)
-- Server:
	player:addItem(ITEM_BAG, 1, false, CONST_SLOT_BACKPACK):setShader("Map - Party")
```

5/6  send disableFeature / enableFeature.  easy to modify in config.lua 

```lua
const auto &enabledFeatures = g_configManager().getEnabledOTCFeatures();
const auto &disabledFeatures = g_configManager().getDisabledOTCFeatures();

for (auto feature : enabledFeatures) {
g_logger().info("Feature enabled: {}", feature);
}

for (auto feature : disabledFeatures) {
g_logger().info("Feature disabled: {}", feature);
}
```

6/6

![image](https://github.com/user-attachments/assets/a13026dd-aba6-48ed-bec7-b466c48a0dfb)
![image](https://github.com/user-attachments/assets/a8ea4b28-e9ab-4d78-a571-a0efab8cf68c)

```lua
x:addCustomOutfit("aura", 8)
x:removeCustomOutfit("aura", 8)
```







```lua
local talkAction = TalkAction("!testOTCR")

local options = {
	Creature = {
		{
			name = "set attach Effect 7",
			action = function(player)
				player:attachEffectById(7)
			end,
		},
		{
			name = "get all attach Effect",
			action = function(player)
				local effects = player:getAttachedEffects()
				if #effects > 0 then
					local effectsString = "your AE are: " .. table.concat(effects, ", ")
					player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, effectsString)
				else
					player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, "you do not have any AE.")
				end
			end,
		},
		{
			name = "detach Effect 7",
			action = function(player)
				player:detachEffectById(7)
			end,
		},
		{
			name = "set Shader player",
			action = function(player)
				player:setShader("Outfit - Rainbow")
			end,
		},
		{
			name = "no shader player",
			action = function(player)
				player:setShader("")
			end,
		},
		{
			name = "get Shader player",
			action = function(player)
				player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, player:getShader())
			end,
		},
	},
	Map = {
		{
			name = "set Shader Map",
			action = function(player)
				player:setMapShader("Map - Party")
			end,
		},
		{
			name = "no shader Map",
			action = function(player)
				player:setMapShader("")
			end,
		},
		{
			name = "get Shader Map",
			action = function(player)
				player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, player:getMapShader())
			end,
		},
	},
	Items = { {
		name = "Shader",
		action = function(player)
			local item = player:addItem(ITEM_BAG, 1, false, CONST_SLOT_BACKPACK)
			item:setShader("Map - Party")
			player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, "item have shader : " .. item:getShader())
		end,
	} },
	game_outfit = {
		{
			name = "Add permanent aura 8",
			action = function(player)
				player:addCustomOutfit("aura", 8)
				player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, "add aura 8")
			end,
		},
		{
			name = "remove permanent aura 8",
			action = function(player)
				player:removeCustomOutfit("aura", 8)
				player:sendTextMessage(MESSAGE_GAMEMASTER_CONSOLE, "remove aura 8")
			end,
		},
	},
}

function talkAction.onSay(player, words, param, type)
	local modalWindow = ModalWindow({
		title = "Fast Test",
		message = "Choose one of the following options:",
	})

	for category, subOptions in pairs(options) do
		modalWindow:addChoice(category, function(player, button, choice)
			local subModalWindow = ModalWindow({
				title = "Select a Sub-Option",
				message = "Choose one of the following sub-options for " .. category .. ":",
			})

			for _, subOption in ipairs(subOptions) do
				subModalWindow:addChoice(subOption.name, function(plyr, btn, ch)
					if button.name == "OK" then
						subOption.action(plyr)
					end
				end)
			end

			subModalWindow:addButton("OK")
			subModalWindow:addButton("Cancel", function(plyr)
				subModalWindow:clear()
			end)
			subModalWindow:sendToPlayer(player)
		end)
	end

	modalWindow:addButton("OK")
	modalWindow:addButton("Cancel", function(player)
		modalWindow:clear()
	end)
	modalWindow:sendToPlayer(player)
	return false
end

talkAction:groupType("normal")
talkAction:separator(" ")
talkAction:register()

local playerLoginTestOTCR = CreatureEvent("simpletest")

function playerLoginTestOTCR.onLogin(player)
	if player:getClient().os >= 10 and player:getClient().os < 12 then
		player:say("say !testOTCR", TALKTYPE_SAY)
	end
	return true
end

playerLoginTestOTCR:register()
```






**Test Configuration**:

  - Server Version: 13.40
  - Client: OTCR
  - Operating System: 11

## Checklist

  - [ ] My code follows the style guidelines of this project
  - [ ] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works
  
### Note: To be honest, I'm not good at C++, and I don't use Canary.

kokekanon committed:
1/6 feature: Creature : AttachedEffect, Shader

attachEffectById
detachEffectById
getShader
setShader

kokekanon committed:
2/7 feature: Map Shader

player:setMapShader("Map - Party")

kokekanon committed:
3/7 feature: TypingIcon

mehah commented on src/game/game.cpp:
```diff
@@ -10875,8 +10875,24 @@ void Game::updateCreatureShader(const std::shared_ptr<Creature> &creature) {
+	}
+	auto spectators = Spectators().find<Creature>(player->getPosition(), true);
+	auto playersSpectators = spectators.filter<Player>();
+	for (const auto &spectator : spectators) {
```
remove
```c++
auto spectators = Spectators().find<Creature>(player->getPosition(), true);
auto playersSpectators = spectators.filter<Player>();
```
replace
```c++
for (const auto &spectator : spectators) {
```
to
```c++
	for (const auto &spectator : Spectators().find<Player>(player->getPosition(), true)) {

mehah commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,47 @@ void Game::updatePlayersOnline() const {
 }
+
+void Game::sendAttachedEffect(const std::shared_ptr<Creature> &creature, uint16_t effectId) {
+	auto spectators = Spectators().find<Creature>(creature->getPosition(), true);
```
`<Creature>` to `<Player>`

mehah commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,47 @@ void Game::updatePlayersOnline() const {
+	}
+}
+void Game::sendDetachEffect(const std::shared_ptr<Creature> &creature, uint16_t effectId) {
+	auto spectators = Spectators().find<Creature>(creature->getPosition(), true);
```
`<Creature>` to `<Player>`

mehah commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,47 @@ void Game::updatePlayersOnline() const {
+	}
+}
+void Game::updateCreatureShader(const std::shared_ptr<Creature> &creature) {
+	auto spectators = Spectators().find<Creature>(creature->getPosition(), true);
```
`<Creature>` to `<Player>`

kokekanon committed:
review: mehah

kokekanon committed:
4/7 feature: Item Shader

kokekanon committed:
5/7 send disableFeature / enableFeature

const auto &enabledFeatures = g_configManager().getEnabledOTCFeatures();
const auto &disabledFeatures = g_configManager().getDisabledOTCFeatures();

				for (int32_t feature : enabledFeatures) {
					g_logger().info("Feature enabled: {}", feature);
				}

				for (int32_t feature : disabledFeatures) {
					g_logger().info("Feature disabled: {}", feature);
				}

Update configmanager.hpp

dudantas commented on src/creatures/creature.cpp:
```diff
@@ -2002,3 +2002,20 @@ void Creature::sendAsyncTasks() {
+	g_game().sendAttachedEffect(static_self_cast<Creature>(), id);
+}
+void Creature::detachEffectById(uint16_t id) {
+	auto it = std::find(attachedEffectList.begin(), attachedEffectList.end(), id);
```
```suggestion

void Creature::detachEffectById(uint16_t id) {
	auto it = std::ranges::find(attachedEffectList, id);
```

dudantas commented on src/creatures/creature.cpp:
```diff
@@ -2002,3 +2002,20 @@ void Creature::sendAsyncTasks() {
 }
+
+void Creature::attachEffectById(uint16_t id) {
+	auto it = std::find(attachedEffectList.begin(), attachedEffectList.end(), id);
```
```suggestion
	auto it = std::ranges::find(attachedEffectList, id);
```

dudantas commented on src/creatures/creature.hpp:
```diff
@@ -692,6 +692,17 @@ class Creature : virtual public Thing, public SharedObject {
+	std::string getShader() const {
+		return shader;
+	}
+	void setShader(const std::string &shaderName) {
```
```suggestion

	void setShader(std::string_view shaderName) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9572,6 +9572,53 @@ void Player::sendLootContainers() const {
+		return;
+	}
+	if (creature->getPlayer()) {
+		if (client) {
```
You don't need to check the "client" as it is already checked at the beginning of the function, it is redundant

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9572,6 +9572,53 @@ void Player::sendLootContainers() const {
+		return;
+	}
+	if (creature->getPlayer()) {
+		if (client) {
```
Same here, you can remove this if statement

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9572,6 +9572,53 @@ void Player::sendLootContainers() const {
+		return;
+	}
+	if (creature->getPlayer()) {
+		if (client) {
```
Same

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9572,6 +9572,53 @@ void Player::sendLootContainers() const {
+}
+
+void Player::sendMapShader(const std::string &shaderName) const {
+	if (client) {
```
Perhaps using the same "return" pattern would be better, this way it standardizes.

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9572,6 +9572,53 @@ void Player::sendLootContainers() const {
+}
+
+void Player::sendPlayerTyping(const std::shared_ptr<Creature> &creature, uint8_t typing) const {
+	if (client) {
```
Same here (above)

dudantas commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,79 @@ void Game::updatePlayersOnline() const {
+void Game::sendAttachedEffect(const std::shared_ptr<Creature> &creature, uint16_t effectId) {
+	auto spectators = Spectators().find<Player>(creature->getPosition(), true);
+	for (const auto &spectator : spectators) {
+		auto player = spectator->getPlayer();
```
```suggestion
		const auto &player = spectator->getPlayer();
```

dudantas commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,79 @@ void Game::updatePlayersOnline() const {
+void Game::sendDetachEffect(const std::shared_ptr<Creature> &creature, uint16_t effectId) {
+	auto spectators = Spectators().find<Player>(creature->getPosition(), true);
+	for (const auto &spectator : spectators) {
+		auto player = spectator->getPlayer();
```
Can be "const ref"

dudantas commented on src/game/game.cpp:
```diff
@@ -10852,3 +10852,79 @@ void Game::updatePlayersOnline() const {
+void Game::updateCreatureShader(const std::shared_ptr<Creature> &creature) {
+	auto spectators = Spectators().find<Player>(creature->getPosition(), true);
+	for (const auto &spectator : spectators) {
+		auto player = spectator->getPlayer();
```
Can be "const ref"

dudantas commented on src/items/item.hpp:
```diff
@@ -141,6 +141,24 @@ class ItemProperties {
 		return getCorpseOwner() == static_cast<uint32_t>(std::numeric_limits<int32_t>::max());
 	}
 
+	void setShader(const std::string &shaderName) {
```
Maybe here can be "string_view"?

kokekanon replied:
it's a newbie question, but how do I use string_view without changing 
msg.addstring, setCustomAttribute, pushString.( are previously defined as string.)

I always get this error
![image](https://github.com/user-attachments/assets/38e58ced-6b42-4458-84c7-abe212797bd1)

I converted to string xd, temporarily to compile without errors.

```cpp
msg.addString(std::string(item->getShader())); // temp
```


![image](https://github.com/user-attachments/assets/bfe9671f-c39c-4d6c-8346-a107f3d3d15e)
![image](https://github.com/user-attachments/assets/155b510d-d2c7-45a0-a6ab-5a4dce4af8dc)
![image](https://github.com/user-attachments/assets/03384aa5-bba0-4e58-8be3-170f25f66c19)

dudantas replied:
> it's a newbie question, but how do I use string_view without changing msg.addstring, setCustomAttribute, pushString.( are previously defined as string.)

In this case, you can ignore, it's preferable to keep it as a const reference. Using std::string_view would require creating a std::string with std::string(stringView), which would result in a similar overhead to passing the string by const reference.

dudantas reviewed (commented):
Thank you for your contribution. I have a few observations that could help further improve the work.

kokekanon committed:
review: dudantas

kokekanon committed:
fix sonarcloud bot

kokekanon committed:
fix sonarcloud bot 2

dudantas commented on src/creatures/creature.hpp:
```diff
@@ -695,7 +695,7 @@ class Creature : virtual public Thing, public SharedObject {
 		return shader;
 	}
-	void setShader(const std::string &shaderName) {
+	void setShader(const std::string_view &shaderName) {
```
You can remove the & as well; passing a std::string_view by reference actually increases overhead. std::string_view is intended to be passed by value since it is simply a lightweight view of the string.

dudantas commented on src/creatures/players/player.hpp:
```diff
@@ -1383,7 +1383,7 @@ class Player final : public Creature, public Cylinder, public Bankable {
 	std::string guildNick;
 	std::string loyaltyTitle;
-	std::string mapShader;
+	std::string_view mapShader;
```
Here, it will remain as std::string. std::string_view is used to avoid copying the string, but in the final context where it needs to be stored or modified, it must be a std::string.

dudantas commented:
Apologies for the confusion. In the setShader method, you can continue using std::string_view, but for all other methods, they can remain as std::string.

PS: If you need, I can help commit.

dudantas committed:
fix: wrong review

github-actions[bot] committed:
Code format - (Clang-format)

dudantas commented:
@kokekanon 
I couldn't commit directly to your fork, but here's the commit if you want to cherry pick it
https://github.com/opentibiabr/canary/commit/1437b7eff2fa52acfd5ae9892bd8899d0d349535

kokekanon committed:
fix: itemShader old protocol otcr

kokekanon committed:
clean variable

kokekanon committed:
fix return get shader

kokekanon committed:
missing content player:getAttachedEffects

kokekanon committed:
simple test Talkations. delete when finished

github-actions[bot] committed:
Lua code format - (Stylua)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
format

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
improve: variable name

kokekanon committed:
Merge branch 'kokekanon/all-feature-redemption' of https://github.com/kokekanon/canary3 into kokekanon/all-feature-redemption

kokekanon committed:
game_outfit functional version, not cleaned or optimized

reload required ?
iOLoginData load, save?
need clean Player.cpp

kokekanon committed:
this requires review canary team .no break v8 and cipsoft

github-actions[bot] committed:
Lua code format - (Stylua)

kokekanon commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -1719,7 +1719,7 @@ void ProtocolGame::parseSetOutfit(NetworkMessage &msg) {
 			}
 
-			uint8_t isMountRandomized = msg.getByte();
+			uint8_t isMountRandomized = !oldProtocol ? msg.getByte() : 0;
```
v8 old protocol no send "isMountRandomized" , last packet  is `newOutfit.lookMount = msg.get<uint16_t>(); ` canary  read in (L1711)

![image](https://github.com/user-attachments/assets/b7043095-f016-464d-926d-e0597f2ac329)
https://github.com/opentibiabr/otcv8/blob/main/src/client/protocolgamesend.cpp#L833

Redemption 11.00 no send isMountRandomized only in 13.40

kokekanon commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -3290,12 +3290,6 @@ void ProtocolGame::sendCreatureOutfit(const std::shared_ptr<Creature> &creature,
-		msg.addByte(newOutfit.lookMountLegs);
-		msg.addByte(newOutfit.lookMountFeet);
-	}
 	writeToOutputBuffer(msg);
```
load this (sendCreatureOutfit and AddCreature) into AddOutfit
 (same packets)
![image](https://github.com/user-attachments/assets/1f96b17a-0a03-404e-bd24-2597f2a6e5e0)

kokekanon commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -7738,12 +7734,6 @@ void ProtocolGame::AddCreature(NetworkMessage &msg, const std::shared_ptr<Creatu
-			msg.addByte(outfit.lookMountLegs);
-			msg.addByte(outfit.lookMountFeet);
-		}
 	} else {
```
load this (sendCreatureOutfit and AddCreature) into AddOutfit
 (same packets)
![image](https://github.com/user-attachments/assets/1f96b17a-0a03-404e-bd24-2597f2a6e5e0)

kokekanon committed:
fix: review sonarcloud ?

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge branch 'kokekanon/all-feature-redemption' of https://github.com/kokekanon/canary3 into kokekanon/all-feature-redemption

kokekanon committed:
simple test

github-actions[bot] committed:
Lua code format - (Stylua)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge branch 'kokekanon/all-feature-redemption' of https://github.com/kokekanon/canary3 into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
murge

kokekanon committed:
fix: compilation windows

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

github-actions[bot] committed:
Lua code format - (Stylua)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge branch 'kokekanon/all-feature-redemption' of https://github.com/kokekanon/canary3 into kokekanon/all-feature-redemption

kokekanon committed:
delete dev

kokekanon committed:
fix: linux

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
fix: warning actions github linux

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

mehah commented on src/creatures/appearance/attachedeffects/attachedeffects.hpp:
```diff
@@ -0,0 +1,73 @@
+	}
+
+private:
+	phmap::parallel_flat_hash_set<std::shared_ptr<Aura>> auras;
```
use std::vector, as there is linear search and unfortunately phmap's flat_hash is extremely slow.

mehah commented on src/creatures/appearance/attachedeffects/attachedeffects.hpp:
```diff
@@ -0,0 +1,73 @@
+	phmap::parallel_flat_hash_set<std::shared_ptr<Aura>> auras;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Shader>> shaders;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Effect>> effects;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Wing>> wings;
```
std::vector

mehah commented on src/creatures/appearance/attachedeffects/attachedeffects.hpp:
```diff
@@ -0,0 +1,73 @@
+private:
+	phmap::parallel_flat_hash_set<std::shared_ptr<Aura>> auras;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Shader>> shaders;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Effect>> effects;
```
std::vector

mehah commented on src/creatures/appearance/attachedeffects/attachedeffects.hpp:
```diff
@@ -0,0 +1,73 @@
+
+private:
+	phmap::parallel_flat_hash_set<std::shared_ptr<Aura>> auras;
+	phmap::parallel_flat_hash_set<std::shared_ptr<Shader>> shaders;
```
std::vector

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
review mehah: use std::vector

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+	if (value > 0) {
+		return value;
+	}
+	return static_cast<uint8_t>(kv()->get("last-wing")->get<int>());
```
std::optional may return nullptr, so you need to validate it before getting **it**, or it can cause a crash

```suggestion
	auto wingOpt = kv()->get("last-wing")
	return static_cast<uint8_t>(wingOpt  ? wingOpt ->get<int>() : 0);
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+		}
+	}
+
+	const auto availableWingsSize = static_cast<int32_t>(availableWings.size() - 1);
```
You need to validate this, same as here: https://github.com/opentibiabr/canary/pull/3236

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+bool Player::untameWing(uint8_t wingId) {
+	if (!g_game().attachedeffects->getWingByID(wingId)) {
```
Set to const ref:
```suggestion
	const auto &wingPtr = g_game().attachedeffects->getWingByID(wingId);
	if (!wingPtr) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+void Player::diswing() {
+	const auto &wing = g_game().attachedeffects->getWingByID(getCurrentWing());
```
Unused? Or missing use?

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+	if (value > 0) {
+		return value;
+	}
+	return static_cast<uint8_t>(kv()->get("last-aura")->get<int>());
```
Same here, need check the nullptr opt

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+
+uint8_t Player::getRandomAuraId() const {
+	std::vector<uint8_t> playerAuras;
+	const auto auras = g_game().attachedeffects->getAuras();
```
Avoid unecessary copy, using const ref: const auto &auras

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+		}
+	}
+
+	const auto playerAurasSize = static_cast<int32_t>(playerAuras.size() - 1);
```
Same here, validate out of range index (or empty)

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+bool Player::tameAura(uint8_t auraId) {
+	if (!g_game().attachedeffects->getAuraByID(auraId)) {
```
```suggestion
	const auto &auraPtr = g_game().attachedeffects->getAuraByID(auraId);
	if (!auraPtr) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+bool Player::untameAura(uint8_t auraId) {
+	if (!g_game().attachedeffects->getAuraByID(auraId)) {
```
Use const ref here

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+void Player::disaura() {
+	const auto &aura = g_game().attachedeffects->getAuraByID(getCurrentAura());
```
Unused or missing use.

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+	if (value > 0) {
+		return value;
+	}
+	return static_cast<uint8_t>(kv()->get("last-effect")->get<int>());
```
Validate nullptr std::optional

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+
+uint8_t Player::getRandomEffectId() const {
+	std::vector<uint8_t> playerEffects;
+	const auto effects = g_game().attachedeffects->getEffects();
```
const auto &

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+		}
+	}
+
+	const auto playerEffectsSize = static_cast<int32_t>(playerEffects.size() - 1);
```
Validate out of range

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+void Player::diseffect() {
+	const auto &effect = g_game().attachedeffects->getEffectByID(getCurrentEffect());
```
Unused

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+bool Player::untameShader(uint16_t shaderId) {
+	auto shaderPtr = g_game().attachedeffects->getShaderByID(shaderId);
```
const auto &

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+	// test proposal
+
+	uint16_t elementId;
+	if (std::holds_alternative<uint16_t>(idOrName)) {
```
If this is used very frequently or continuously, it is advisable to avoid using holds alternative, it adds significant overhead as it needs to evaluate the value at runtime.

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+		const std::string &name = std::get<std::string>(idOrName);
+
+		if (type == "wing") {
+			auto element = g_game().attachedeffects->getWingByName(name);
```
const auto & (and the others same, aura, shader, etc)

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+		const std::string &name = std::get<std::string>(idOrName);
+
+		if (type == "wings") {
+			auto element = g_game().attachedeffects->getWingByName(name);
```
const auto &

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -9630,6 +10399,52 @@ void Player::sendLootContainers() const {
+		return;
+	}
+
+	if (creature->getPlayer()) {
```
If this is only executed for player, it is better to change this to stay in the player class and change the default to "Player" instead of creature (same for the others)

dudantas commented on src/creatures/players/player.hpp:
```diff
@@ -198,6 +203,70 @@ class Player final : public Creature, public Cylinder, public Bankable {
 	bool hasAnyMount() const;
 	uint8_t getRandomMountId() const;
 	void dismount();
+
```
Perhaps it would be better to create a scoped object for this, similar to what's done with the cyclopedia, player VIP, player titles, etc. This way, we can avoid adding even more variables to the Player class, which is already quite large. For example, OtClientFeatures class?

https://github.com/opentibiabr/canary/blob/main/src/creatures/players/vip/player_vip.cpp
https://github.com/opentibiabr/canary/blob/3ac0bab131658df70151f2c4bfb87564cb213cf3/src/creatures/players/player.hpp#L1272
https://github.com/opentibiabr/canary/blob/3ac0bab131658df70151f2c4bfb87564cb213cf3/src/creatures/players/player.cpp#L75

dudantas commented on src/creatures/players/player.hpp:
```diff
@@ -1374,6 +1455,10 @@ class Player final : public Creature, public Cylinder, public Bankable {
 	std::vector<OutfitEntry> outfits;
+	std::unordered_set<uint16_t> wings;
+	std::unordered_set<uint16_t> auras;
+	std::unordered_set<uint16_t> effects;
```
Unused

dudantas commented on src/game/game.cpp:
```diff
@@ -6189,6 +6192,82 @@ void Game::playerChangeOutfit(uint32_t playerId, Outfit_t outfit, uint8_t isMoun
+
+	// @  wings
+	if (outfit.lookWing != 0) {
+		const auto wing = attachedeffects->getWingByID(outfit.lookWing);
```
const auto &

dudantas commented on src/game/game.cpp:
```diff
@@ -6189,6 +6192,82 @@ void Game::playerChangeOutfit(uint32_t playerId, Outfit_t outfit, uint8_t isMoun
+	// @
+	// @  Effect
+	if (outfit.lookEffect != 0) {
+		const auto effect = attachedeffects->getEffectByID(outfit.lookEffect);
```
const auto &

dudantas commented on src/game/game.cpp:
```diff
@@ -6189,6 +6192,82 @@ void Game::playerChangeOutfit(uint32_t playerId, Outfit_t outfit, uint8_t isMoun
+	// @
+	// @  Aura
+	if (outfit.lookAura != 0) {
+		const auto aura = attachedeffects->getAuraByID(outfit.lookAura);
```
const auto &

dudantas commented on src/game/game.cpp:
```diff
@@ -6189,6 +6192,82 @@ void Game::playerChangeOutfit(uint32_t playerId, Outfit_t outfit, uint8_t isMoun
+	// @
+	/// shaders
+	if (outfit.lookShader != 0) {
+		const auto shaderPtr = attachedeffects->getShaderByID(outfit.lookShader);
```
const auto &

dudantas commented on src/game/game.cpp:
```diff
@@ -10985,6 +11064,83 @@ void Game::updatePlayersOnline() const {
+}
+
+void Game::refreshItem(const std::shared_ptr<Item> &item) {
+	if (!item || !item->getParent()) {
```
You can remove the "getParent"

dudantas commented on src/game/game.cpp:
```diff
@@ -10985,6 +11064,83 @@ void Game::updatePlayersOnline() const {
+	if (!item || !item->getParent()) {
+		return;
+	}
+	const auto &parent = item->getParent();
```
```suggestion
	const auto &parent = item->getParent();
	if (!parent) {
		return;
	}
```

dudantas commented on src/game/game.cpp:
```diff
@@ -10985,6 +11064,83 @@ void Game::updatePlayersOnline() const {
+			const auto spectators = Spectators().find<Player>(container->getPosition(), false, 2, 2, 2, 2);
+			// send to client
+			for (const auto &spectator : spectators) {
+				spectator->getPlayer()->sendUpdateContainerItem(container, static_cast<uint16_t>(index), item);
```
Since it is executed in a loop, it is interesting to extract getPlayer into a const ref variable to avoid incrementing the reference counter.
```suggestion
				const auto &player = spectator->getPlayer();
				if (!player) {
					continue;
				}
				player->sendUpdateContainerItem(container, static_cast<uint16_t>(index), item);
```

dudantas commented on src/game/game.cpp:
```diff
@@ -10985,6 +11064,83 @@ void Game::updatePlayersOnline() const {
+		const auto spectators = Spectators().find<Player>(tile->getPosition(), true);
+		// send to client
+		for (const auto &spectator : spectators) {
+			spectator->getPlayer()->sendUpdateTileItem(tile, tile->getPosition(), item);
```
Same here

dudantas commented on src/lua/functions/creatures/creature_functions.cpp:
```diff
@@ -1186,3 +1191,80 @@ int CreatureFunctions::luaCreatureClearIcons(lua_State* L) {
+	const auto &creature = Lua::getUserdataShared<Creature>(L, 1);
+	if (!creature) {
+		Lua::reportErrorFunc(Lua::getErrorDesc(LUA_ERROR_CREATURE_NOT_FOUND));
+		Lua::pushBoolean(L, false);
```
You just need to use reportErrorFunc, it already pushes the string with the error, so pushing bool or nil is redundant. (Same for the others)

dudantas commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9354,6 +9447,155 @@ void ProtocolGame::sendTakeScreenshot(Screenshot_t screenshotType) {
+	uint16_t limitWings = std::numeric_limits<uint16_t>::max();
+	uint16_t wingSize = 0;
+	msg.skipBytes(1);
+	const auto wings = g_game().attachedeffects->getWings();
```
const auto &

dudantas commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9354,6 +9447,155 @@ void ProtocolGame::sendTakeScreenshot(Screenshot_t screenshotType) {
+	uint16_t limitAuras = std::numeric_limits<uint16_t>::max();
+	uint16_t auraSize = 0;
+	msg.skipBytes(1);
+	const auto auras = g_game().attachedeffects->getAuras();
```
const auto &

dudantas commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9354,6 +9447,155 @@ void ProtocolGame::sendTakeScreenshot(Screenshot_t screenshotType) {
+	uint16_t limitEffects = std::numeric_limits<uint16_t>::max();
+	uint16_t effectSize = 0;
+	msg.skipBytes(1);
+	const auto effects = g_game().attachedeffects->getEffects();
```
const auto &

dudantas reviewed (commented):
Good work. I have some small observations that may impact performance/stability.

kokekanon committed:
review: dudantas part 1/2

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
fix

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
fix: crash in last appearance attachedeffects on debug compilation 3284

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
add comments

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
1>D:\desktop\canary3\src\lua\functions\creatures\player\player_functions.cpp(4993): warning C4715: 'PlayerFunctions::luaPlayerRemoveCustomOutfit': not all control paths

1>D:\desktop\canary3\src\lua\functions\creatures\player\player_functions.cpp(4993): warning C4715: 'PlayerFunctions::luaPlayerRemoveCustomOutfit': not all control paths

kokekanon committed:
Temp to avoid a bug in OTC if the "OTCRFeatures" array is not declared in config.lua.

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
fix: post murge

dudantas committed:
Merge branch 'main' into kokekanon/all-feature-redemption

dudantas committed:
improve: extract codes from player class to scoped object

github-actions[bot] committed:
Code format - (Clang-format)

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

kokekanon committed:
fix: compilation vs solution

kokekanon committed:
format

kokekanon committed:
format

vllsystems reviewed (approved):
![image](https://github.com/user-attachments/assets/56f55754-bb4f-4dbb-9888-753d3804df8d)
Working fine.

kokekanon committed:
Merge remote-tracking branch 'upstream/main' into kokekanon/all-feature-redemption

cmcpro7 commented:
From what I noticed, I can only log in using this fix in OTCv8, if I log in using the official client I can't.

cmcpro7 commented:
![Captura de tela 2025-03-24 174644](https://github.com/user-attachments/assets/8518bed8-34c7-4d54-934f-a85938baf38c)

dudantas commented:
> From what I noticed, I can only log in using this fix in OTCv8, if I log in using the official client I can't.

This PR don't change anything from official client.

cmcpro7 commented:
> > Pelo que percebi, só consigo efetuar login usando essa correção no OTCv8. Se eu efetuar login usando o cliente oficial, não consigo.
> 
> Este RP não altera nada em relação ao cliente oficial.

Yes, I ended up expressing myself wrong. I say that I can only access it through the official client. When I try through OTCv8 it doesn't log in, it gives an error.

kokekanon commented:
> Yes, I ended up expressing myself wrong. I say that I can only access it through the official client. When I try through OTCv8 it doesn't log in, it gives an error.

main repo

https://github.com/user-attachments/assets/a27403a3-d8b9-41aa-9933-6b0e9c16c3b8

the only packet error you see in the video is not caused by this pr. 
I think it is because of the cyclopedia/house pr, it should be a conditional if (oldprotocol) {} in `src\server\network\protocol\protocolgame.cpp.`

you can also be fixed client side, but the v8 repo kondra is dead and in OTA I don't think they will accept my pr with the solution 

-------
and in this pr , there is no modification to the outputmessage header. error of your image.

cmcpro7 commented:
> > Yes, I ended up expressing myself wrong. I say that I can only access it through the official client. When I try through OTCv8 it doesn't log in, it gives an error.
> 
> main repo
> 
>  2025-04-04.10-11-16.mp4 
> the only packet error you see in the video is not caused by this pr. I think it is because of the cyclopedia/house pr, it should be a conditional if (oldprotocol) {}.
> 
> you can also be fixed client side, but the v8 repo kondra is dead and in OTA I don't think they will accept my pr with the solution
> 
> and in this pr , there is no modification to the outputmessage header. error of your image.



I understand my friend, so basically it would be an error regarding the cyclopedia house on the otcv8 side?
I think I would have to adjust this part in the otcv8 source, correct?
